### DISCUSSION TITLE
CSRF token mismatch while testing

### DISCUSSION THREAD

**Discussion by maan56** [laravel/framework#40467] [Q&A] [answered]
Hi there,
I'm testing my application and I have created tests for a post route and expected validation errors and for other test expected validation passes.
one of the tests which expects validation passes, just fine. but the other one failed with error "Session is missing expected key [errors]." And when I dump the response, it shows "CSRF token mismatch" Error with error code 419.
Here comes the headache!
I always run `php artisan optimize` after updating routes or config files. I was searching on google and I found out one solution. So I cleared configuration cache with `php artisan config:clear` and run the test again, and it is all green.

So, my question is without clearing configuration cache why did one test passes and the other one fail on the same post route?

**Comment by SjorsO** [comment, marked as answer]
CSRF protection is skipped when running tests. That happens in the [VerivyCsrfToken middleware](https://github.com/laravel/framework/blob/9156c997547922cce17dbbbddd8b2efb22312f68/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php#L74). Laravel knows you're running tests if your APP_ENV is set to `testing` (which is set automatically in [phpunit.xml](https://github.com/laravel/laravel/blob/8d08717f15d1bddf3d28124b928ccc3a5cfdf84b/phpunit.xml#L21)). When you cache your config file, your APP_ENV is hardcoded to whatever is in your `.env`file at the time (probably `local`). Then when you run your tests, Laravel thinks you're not running tests, and will check the CSRF token.

As a general tip, never run `artisan optimize` or `config:cache` in development. Only use them in production.

**Reply by maan56** [reply]
Got it! So now I understand why my other test was passing. Because I was only asserting values as valid as `$response->assertValid(['field1']);` and phpunit was looking for errors for 'field1' in session. The session is empty because of configuration values are set to use local environment.

Thanks buddy!

**Comment by Pansiere** [comment]
I resolved this by running the following command:
````PHP
php artisan config:clear 
````
